src/b64.mli

```diff
@@ -24,23 +26,36 @@
 
     {e Release %%VERSION%% - %%PKG_HOMEPAGE%%} *)
 
-val default_alphabet : string
-(** A 64-character string specifying the regular Base64 alphabet. *)
+type alphabet
 
-val uri_safe_alphabet : string
-(** A 64-character string specifying the URI- and filename-safe Base64
+val default_alphabet : alphabet
+(** A 64-character alphabet specifying the regular Base64 alphabet. *)
+
+val uri_safe_alphabet : alphabet
+(** A 64-character alphabet specifying the URI- and filename-safe Base64
     alphabet. *)
 
-val decode : ?alphabet:string -> string -> string
+val make_alphabet : string -> alphabet
+val length_alphabet : alphabet -> int
+
+val decode_exn : ?pad:bool -> ?alphabet:alphabet -> string -> string
 (** [decode s] decodes the string [s] that is encoded in Base64 format. Will
     leave trailing NULLs on the string, padding it out to a multiple of 3
-    characters. [alphabet] defaults to {!default_alphabet}. @raise Not_found if
-    [s] is not a valid Base64 string. *)
+    characters. [alphabet] defaults to {!default_alphabet}.
+
+    Decoder can fail when character of [s] is not a part of [alphabet] or is not
+    [padding] character. If input is not padded correctly, decoder does the
+    best-effort but it does not ensure [decode_exn (encode ~pad:false x) = x].
 
-val decode_opt : ?alphabet:string -> string -> string option
+    @raise if Invalid_argument [s] is not a valid Base64 string. *)
+
+val decode_opt : ?pad:bool -> ?alphabet:alphabet -> string -> string option
```

Comment:
I think we can get just rid of `decode_opt`. It's strictly less expressive than `decode` -- or do you think it's useful with some Option_monad somewhere?